src/darwin/Framework/CHIP/MTROTAUnsolicitedBDXMessageHandler.mm

```diff
@@ -0,0 +1,142 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTROTAUnsolicitedBDXMessageHandler.h"
+
+#include <protocols/bdx/BdxMessages.h>
+
+using namespace chip;
+using namespace chip::Messaging;
+using namespace chip::bdx;
+
+MTROTAUnsolicitedBDXMessageHandler * MTROTAUnsolicitedBDXMessageHandler::sInstance = nullptr;
+
+CHIP_ERROR MTROTAUnsolicitedBDXMessageHandler::Init(ExchangeManager * exchangeManager)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    MTROTAUnsolicitedBDXMessageHandler::mNumberOfDelegates = 0;
+    mExchangeMgr = exchangeManager;
+    return mExchangeMgr->RegisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id, this);
+}
+
+MTROTAUnsolicitedBDXMessageHandler * MTROTAUnsolicitedBDXMessageHandler::GetInstance()
+{
+    return sInstance;
+}
+
+void MTROTAUnsolicitedBDXMessageHandler::Shutdown()
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mExchangeMgr != nullptr);
+
+    mExchangeMgr->UnregisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id);
+    MTROTAUnsolicitedBDXMessageHandler::mNumberOfDelegates = 0;
+
+    if (mOTAImageTransferHandler != nullptr) {
+        mOTAImageTransferHandler->DestroySelf();
+        delete mOTAImageTransferHandler;
+    }
+}
+
+void MTROTAUnsolicitedBDXMessageHandler::ControllerShuttingDown(MTRDeviceController * controller)
+{
+    assertChipStackLockedByCurrentThread();
+
+    // Since the OTA provider delegate only supports one BDX transfer at a time, calling ShutDown is fine for now.
+    // TODO: #36181 - This class needs to keep a list of all MTROTAImageTransferHandlers mapped to the fabric index and only
+    // delete the MTROTAImageTransferHandler objects with a fabric index matching the MTRDeviceController's fabric index.
+    Shutdown();
+}
+
+CHIP_ERROR MTROTAUnsolicitedBDXMessageHandler::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, const SessionHandle & session,
+    ExchangeDelegate * _Nonnull & newDelegate)
+{
+    assertChipStackLockedByCurrentThread();
+
+    ChipLogDetail(BDX, "MTROTAUnsolicitedBDXMessageHandler: OnUnsolicitedMessageReceived: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId,
+        payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+
+    VerifyOrReturnError(mExchangeMgr != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    if (GetNumberOfDelegates() >= 1) {
+        return CHIP_ERROR_BUSY;
```

Comment:
So... this is not responding with a BUSY status on the exchange or anything.  It's going to just drop everything on the floor.  Is this actually the behavior we want?  And if it is, why do we have the higher-level "busy" handling above that sends nice cluster response commands?  This seems pretty broken to me....

Reply:
hmm let me figure out how to pipe the busy correctly. the tests pass where we test busy so either the tests don't test the right thing or something is sending the right response but cleaning up the class which it shouldn't.

Reply:
do a follow up. create multiple handlers.

Reply:
#36526. combined in the follow up issue